### ISSUE TITLE
Design CDK for Terraform integration testing

### ISSUE THREAD

**Issue by danieldreier** [hashicorp/terraform-cdk#1093]
This is a research and design task to design and plan a workflow for integration testing of CDK for Terraform applications.  This ticket is a placeholder for the sake of long-term planning, and should be rewritten with more defined scope before work starts.

The scope of this issue is to propose an approach to integration testing, ideally aligned with the AWS CDK approach linked below. An internal HashiCorp RFC is needed for this issue before implementation work starts.

Reference Material
* https://github.com/aws/aws-cdk-rfcs/pull/378

**Comment by jokerbaba** [comment]
Lol

**Comment by JohnCalhoun** [comment]
need this feature and would love to help provider feedback or ideas

**Comment by skorfmann** [comment]
> need this feature and would love to help provider feedback or ideas

Which use-cases do you have in mind in particular?

**Comment by skorfmann** [comment]
@marcus-sa 
> Any consideration regarding adding support for something like https://www.terraform.io/docs/language/modules/testing-experiment.html ?

No, not so far - but curious to learn how you'd use this in the context of cdktf

**Comment by marcus-sa** [comment]
> No, not so far - but curious to learn how you'd use this in the context of cdktf

@skorfmann I've done something like this instead.

```ts
/// <reference types="@types/jest" />
import { promises as fs } from 'fs';
import * as path from 'path';
import * as os from 'os';

import { App, TerraformStack, Testing } from 'cdktf';
import { $ } from 'zx';

export interface TerraformTestOptions {
  readonly id: string;
  readonly pluginDir: string;
}

export class TerraformTest {
  private app!: App;
  private stack!: TerraformStack;

  private applied = false;
  private initialized = false;
  private destroyed = false;
  private stackDir!: string;

  static fullSynth(fn: (stack: TerraformStack) => void): string {
    const app = Testing.app();
    const stack = new TerraformStack(app, 'test');

    fn(stack);

    return Testing.fullSynth(stack);
  }

  constructor(private readonly options: TerraformTestOptions) {
    beforeAll(async () => {
      const outDir = await fs.mkdtemp(path.join(os.tmpdir(), 'cdktf.outdir.'));
      this.app = new App({ outdir: outDir });
      this.stackDir = `${outDir}/stacks/${this.options.id}`;
    });

    afterAll(async () => {
      try {
        if (this.initialized) {
          await $`terraform -chdir=${this.stackDir} destroy -input=false -auto-approve -lock=false`;
          this.destroyed = true;
        }
      } finally {
        await fs.rm(this.app.outdir, { recursive: true });
      }
    });
  }

  // TODO: Return outputs
  async synthInitApply(fn: (stack: TerraformStack) => void): Promise<void> {
    this.stack = new TerraformStack(this.app, this.options.id);
    fn(this.stack);

    this.app.synth();

    await $`terraform -chdir=${this.stackDir} init -backend=true -upgrade=false`;
    this.initialized = true;

    await $`terraform -chdir=${this.stackDir} apply -input=false -auto-approve`;
    this.applied = true;
  }
}
```

**Comment by JohnCalhoun** [comment]
yeah the overriding of TerraformVariable is super needed for this. But right now I can test up and including `terraform plan` using to toPlanSuccessfully(), but want i want to do next is actualy launch the stack (from the .test.ts jest file) and then get outputs from the stack and run AWS queries against the stack (like invoke a lambda function and check its return value).

I built something sort of like this at my previous job but I want to do it this time in a more community supported way, I dont want to base our whole codebase testing on a hack that the community will do better anyways.

**Comment by marcus-sa** [comment]
> yeah the overriding of TerraformVariable is super needed for this. But right now I can test up and including `terraform plan` using to toPlanSuccessfully(), but want i want to do next is actualy launch the stack (from the .test.ts jest file) and then get outputs from the stack and run AWS queries against the stack (like invoke a lambda function and check its return value).
> 
> I built something sort of like this at my previous job but I want to do it this time in a more community supported way, I dont want to base our whole codebase testing on a hack that the community will do better anyways.

If you refer to the `terraform.ts` file in the gist I provided, then that can be used to synthesize, initialize and apply a stack, and upon finalizing it'll destroy the stack.

Idk about AWS, but for GCP there's NodeJS clients for all their services that would allow you to do some end-to-end testing.

**Comment by so0k** [comment]
This CDKTF library leverages all the utility functions of terratest for its integration testing.. it does require a bridge to synth the ts into `tf.json`

https://github.com/envtio/base/tree/main/integ
